Libraries/LibWeb/HTML/CanvasRenderingContext2D.cpp

```diff
@@ -61,6 +63,44 @@ void CanvasRenderingContext2D::visit_edges(Cell::Visitor& visitor)
     visitor.visit(m_element);
 }
 
+// https://html.spec.whatwg.org/multipage/canvas.html#canvasrenderingcontext2dsettings
+JS::ThrowCompletionOr<CanvasRenderingContext2DSettings> CanvasRenderingContext2D::context_attributes_from_options(JS::VM& vm, JS::Value value)
+{
+    if (!value.is_nullish() && !value.is_object())
+        return vm.throw_completion<JS::TypeError>(JS::ErrorType::NotAnObjectOfType, "CanvasRenderingContext2DSettings");
+
+    CanvasRenderingContext2DSettings settings;
+    if (value.is_nullish())
+        return settings;
+
+    auto& value_object = value.as_object();
+
+    JS::Value alpha = TRY(value_object.get("alpha"_fly_string));
+    settings.alpha = alpha.is_undefined() ? true : alpha.to_boolean();
+
+    JS::Value desynchronized = TRY(value_object.get("desynchronized"_fly_string));
+    settings.desynchronized = desynchronized.is_undefined() ? false : desynchronized.to_boolean();
+
+    JS::Value color_space = TRY(value_object.get("colorSpace"_fly_string));
+    auto color_space_string = TRY(color_space.to_string(vm));
+    if (color_space_string == "srgb"sv)
+        settings.color_space = Bindings::PredefinedColorSpace::Srgb;
+    else if (color_space_string == "display-p3"sv)
```

Comment:
I think (?) we should be throwing an exception for a TypeError if the enum doesn't match anything. It's a bit unfortunate that we don't have a nice way for the IDL generator to generate this type of conversion for us so that we don't need to craft it ourselves for all of the little quirks that IDL has

Reply:
I can try and take a look at changing the IDL generator to support this and leave that all for later (hopefully not too hairy!) there is a similar type of place we currently need to do this in Streams

Reply:
Yeah, I think we just need to factor out the existing conversion IDL generator can already do into standalone conversion functions instead of inlining everything into the prototype implementation. Then it should be simple to add an attribute to a dictionary to ask the generator to generate that conversion function even if it doesn't appear in any prototype. But touching the IDL generator is always a fun experience 😅